debug RF rx buffers

wildwiz opened the pull request:
Hi, not really a pull request, more a way to share some ideas and possibly get some feedback. I post it here as maybe it is easier to see the changes and make quickly an idea of the stuff. If it is inappropriate just let me know and/or divert the content.

This is a patch for the RFB_DIRECT mode of Sonoff Bridge, that allows to analyze codes that rc-switch fails to decode as long as the rc-switch library grabs them in some way in its buffer. 

The rc-switch library has been modified to instrument an hook to be called whenever it has received anything worthwhile to attempt decode (see change of library in platformio.ini)

On espurna side, I have added a setting "rfb-dump" that wakes up the functionality. Once enabled, to disable it you must reboot the board (the setting is removed as soon as executed the first time, to prevent the board to lock-up in noisy RF environment).

Here is an example of the output (the numbers in the buffer are the microseconds between signal transitions extracted from rc-switch buffer)

The first dump is for one of my RC that I use currently with the RFB_DIRECT patch and that the vanilla Sonoff Bridge fails to recognize.

The second dump is from an RF Thermometer/Hygrometer I am tinkering with.
```
telnet 10.0.0.170
Trying 10.0.0.170...
Connected to 10.0.0.170.
Escape character is '^]'.
[079411] [TELNET] Client #0 connected
set rfb-dump 1
+OK

[103131] [RFBRIDGE] RF buffer count: 50, data:  11059 375 1050 376 1054 373 1053 1088 339 369 1056 1081 348 363 1063 1075 352 361 1066 361 1065 363 1064 1071 356 356 1072 1069 360 351 1075 1064 365 347 1079 1060 370 341 1085 343 1086 1055 374 1049 384 1039 395 1029 415 293
[103249] [RFBRIDGE] Received code: 0015154F
[103250] [RFBRIDGE] Action 0xA4
[103250] [RFBRIDGE] Forward message 'C0010000180015154F'

[350343] [RFBRIDGE] RF buffer count: 74, data:  9122 645 1847 599 3789 601 1846 597 3807 585 3804 601 3787 602 3802 603 1829 600 3803 601 3787 602 1845 601 3787 602 3802 604 1827 600 1846 601 1846 598 1833 596 1852 595 1851 593 1838 592 3811 593 3796 595 3808 596 1835 594 1852 594 1853 593 1838 592 1854 589 1858 589 1842 587 3816 589 1842 590 3813 591 1856 574 3814 592 1856 590
```

xoseperez commented:
It a good idea. Why not making it a setting in the web interface? something you can enable/disable and read the output in the webUI debug panel. Using the same logic it will always boot disabled.

Is there any change your branch for the rc-switch library could be merged to the main code?

wildwiz commented:
* I have added a define rc-switch side and a check ESPurna side so that the firmware can dynamically enable the hook when available. Now you both `set rfb-dump 1` and back `set rfb-dump 0`, the first received RF switches the mode and wipes the setting.

* The output that is available on the telnet line is already available in the webUI too, so if you send a `set rfb-dump 1` or `set rfb-dump 0`  via the debug page, you can see the results there. However, the behavior of the webUI console is a bit odd. Sometimes, multiple lines get collapsed into one, I am not sure why that happens. Maybe too overconfidence in using the DEBUG_xxx calls ?

* About adding a button, after sorting out this issue, I could do ... just, how does that fit with the idea of having the setting removed to prevent crash loops ?

* About the chance of the rc-switch patch being integrated ... I will give it a try.

xoseperez commented:
Yes, the webUI debug panel is not 100% responsible, it won't cope with multiple messages sent very quickly. Actually it has an overflow mechanism to prevent them from making it run out of memory so it will discard messages...

wildwiz commented:
Changes: 
* added a couple of buttons, `Trace RF` and `Untrace RF` on the debug page
* added RF dump quenching limiting to 1 RF dump per second, the web UI seems to work better
* added auto turn-off after 15 minutes

Modes:
* set `rfb-dump 0` -- disables RF DUMP
* set `rfb-dump 1` -- enables RF DUMP with message quenching to 1 dump per second
* set `rfb-dump 2` -- enables integral RF DUMP with no quenching

![debug panel](https://imagebin.ca/v/42LiUSbHzKR2)

wildwiz commented:
A prebuilt TEST image is available here  
https://raw.githubusercontent.com/wiki/wildwiz/espurna/files/espurna-1.12.7a-itead-sonoff-rfbridge-direct-diagnostic-ota.bin

derFrickler commented:
thanks a lot! will test it as soon as i get some time.

derFrickler commented:
Just tested the diagnostic image. With my working intertechno remotes i receive a lot of output lines with Trace RF enabled, ending up with sending the received code to the mqtt broker. 
With the non-working remote this is all i receive on one keypress:
`[096213] [RFBRIDGE] RF buffer count: 84, data:  5600 2474 1621 312 274 708 674 316 672 314 669 316 270 715 276 707 673 315 273 709 675 311 274 709 278 708 676 311 274 709 676 313 276 703 678 312 673 312 279 706 279 704 282 705 282 700 283 703 678 307 678 307 681 307 280 702 282 705 282 702 285`
No newline after receiving the code. A 2nd keypress on the remote just appends to this line. so like this:
`[322665] [RFBRIDGE] RF buffer count: 84, data:  5600 2472 1626 314 266 716 670 314 670 318 671 315 271 711 273 709 676 315 271 711 668 323 267 713 275 710 673 313 274 709 676 312 276 708 675 310 674 316 273 709 277 706 280 704 280 703 284 704 679 306 679 305 680 310 276 705 282 702 284 700 286[323692] [RFBRIDGE] RF buffer count: 84, data:  5606 2462 1635 304 280 702 682 306 679 307 679 309 278 704 280 704 680 308 279 701 686 301 282 704 284 698 686 303 282 703 682 305 286 699 685 304 681 305 281 702 286 696 290 696 290 693 291 694 690 297 687 302 683 305 281 699 288 698 285 696 291`

seems that there is nothing triggering the end of receive and processing.
